tests/client/test_call_graph.cpp

```diff
@@ -0,0 +1,233 @@
+            << "Actual calls to root: " << actual.second << std::endl;
+    }
+
+    std::queue<std::shared_ptr<CallGraph::TreeNode>> actual_fnc_queue;
+    std::queue<std::shared_ptr<CallGraph::TreeNode>> expected_fnc_queue;
+    actual_fnc_queue.push(actual.first); 
+    expected_fnc_queue.push(expected.first);
+
+    while (!actual_fnc_queue.empty())
+    {
+        auto& actual_fnc = actual_fnc_queue.front();
+        auto& expected_fnc = expected_fnc_queue.front();
+
+        if (expected_fnc->data.label != actual_fnc->data.label)
+        {
+            return ::testing::AssertionFailure() << std::endl 
+                << "Expected label: " << expected_fnc->data.label << std::endl
+                << "Actual label: " << actual_fnc->data.label << std::endl;
+
+        }
+        if (expected_fnc->data.start_ts != actual_fnc->data.start_ts)
+        {
+            return ::testing::AssertionFailure() << std::endl 
+                << "For node with label: " << expected_fnc->data.label << std::endl
+                << "Expected start_ts: " << expected_fnc->data.start_ts << std::endl
+                << "Actual start_ts: " << actual_fnc->data.start_ts << std::endl;
+        }
+        if (expected_fnc->data.stop_ts  != actual_fnc->data.stop_ts)
+        {
+            return ::testing::AssertionFailure() << std::endl 
+                << "For node with label: " << expected_fnc->data.label << std::endl
+                << "Expected stop_ts: " << expected_fnc->data.stop_ts << std::endl
+                << "Actual stop_ts: " << actual_fnc->data.stop_ts << std::endl;
+        }
+        
+        if (expected_fnc->children.size() != actual_fnc->children.size())
+        {
+            return ::testing::AssertionFailure() << std::endl 
+                << "For node with label: " << expected_fnc->data.label << std::endl
+                << "Expected number of children: " << expected_fnc->children.size() << std::endl
+                << "Actual number of children: " << actual_fnc->children.size() << std::endl;
+        }
+
+        for (size_t i = 0; i < actual_fnc->children.size(); ++i)
+        {
+            if (expected_fnc->children[i].second != actual_fnc->children[i].second) 
+            {
+                return ::testing::AssertionFailure() << std::endl
+                    << "For node with label: " << expected_fnc->data.label << std::endl
+                    << "Expected number of calls to a child: " << expected_fnc->children[i].second << std::endl
+                    << "Actual number of calls to a child: " << actual_fnc->children[i].second << std::endl;
+            }
+
+            actual_fnc_queue.push(actual_fnc->children[i].first);
+            expected_fnc_queue.push(expected_fnc->children[i].first);
+        }
+
+        actual_fnc_queue.pop();
+        expected_fnc_queue.pop();
+    }
+
+    return ::testing::AssertionSuccess();
+}
+
+void init(std::vector<CallGraph::NodeData>& events,
```

Comment:
I think you can just return the tree. It shouldn't have any performance impact (see https://en.cppreference.com/w/cpp/language/copy_elision)

Reply:
This method doesn't just initialize the tree. It also reads the sequence of events which we pass to the CallGraph. So it would have to return two structures and that's why I've chosen to keep references to them.